Forge.TreeWalker/src/ForgeDictionary.cs

```diff
@@ -41,12 +36,12 @@ public class ForgeDictionary : IForgeDictionary
         /// ForgeDictionary Constructor.
         /// </summary>
         /// <param name="forgeStateTable">The forge state table dictionary object.</param>
+        /// <param name="rootSessionId">The unique identifier for the root/parent session.</param>
         /// <param name="sessionId">The unique identifier for this session.</param>
-        public ForgeDictionary(IDictionary<string, object> forgeStateTable, Guid sessionId)
+        public ForgeDictionary(IDictionary<string, object> forgeStateTable, Guid rootSessionId, Guid sessionId)
```

Comment:
>Guid rootSessionId, Guid sessionId [](start = 76, length = 34)

do we need three tuple, rootSessionId (root tree), parentSessionId (for the caller of the sub-tree), and sessionId (for this sub-tree)?

scenario is TreeRoot calls TreeBranchA and TreeBranchB, while A and B can both call the subTreeC. With the two of rootSessionId and SessionId of subTreeC, we cannot tell either BranchA or BranchB calles the subTreeC, until we walk the whole sesssion from RootTree start.

Reply:
The concept of state separation for different sessions works a bit differently than you describe. The subroutine sessions don't need to call back or know about their parent sessions.
The RootSessionId is passed along just so the application can keep track that all the Subroutine sessions belong to one uber-session.
Technically, the RootSessionId is not required to attain state separation.

The SubroutineAction creates a new Guid SessionId for its Subroutine TreeWalkerSession, and persists it in its Intermediates.
Let's say a Subroutine session contains another nested Subroutine session. Everything works the same, a new Guid SessionId is created.
The SessionId alone is enough to uniquely identify any session. You don't have to know anything about its parents.

---
In reply to: [356766078](https://github.com/microsoft/Forge/pull/6#discussion_r356766078) [](ancestors = 356766078)